### DISCUSSION TITLE
DAO contract

### DISCUSSION THREAD

**Discussion by rahimklaber** [stellar/sorobanathon#28] [General]
Hi, I made an example DAO contract. I also tried to explain the code but I'm not that good at writing. Hopefully it helps someone though.

repo : https://github.com/rahimklaber/soroban_dao

**Comment by achallagundla** [comment]
Hey hey rahim, thank you for your submission...will get back to you shortly after review!

**Comment by iamjayrome** [comment]
This is awesome! Your explanations really helped me understand what exactly was going on. Great job.

**Comment by leighmcculloch** [comment]
I really enjoyed reading this contract and the explanation for how it worked. I also found the way the code was structured – with the helper functions at the bottom of the file – made it quite easy to understand from a high level downwards when reading for the first time.  In terms of functionality, I like the bootstrap model. Also the deadline for proposals as well. And the tests made it really to understand how it worked too. The fact that proposals are sets of instructions is simple and elegant.

Some small things I noticed:
 - The admin currently receives 1 share automatically, but this might not be necessary given the admin could give themselves 1 share. 
 - The bootstrap period is controlled by a timestamp, but it could also be a function the admin calls to change a boolean value to indicate bootstrap is finished. The advantage of a function that flips a boolean is less code on chain so cheaper to use. And the DAO could be setup and be usable immediately without waiting a week.
 - The `Proposal` type gets reloaded on every vote. You could reduce the resources required to vote by storing `tot_votes` in a separate data entry since votes really only needs to read/write that one thing (it doesn't need to validate anything). That has some run-on effects: you won't need to check if `tot_votes` is zero in `c_prop`.
 - It might be worth exploring if using errors would make it easier to know what cause errors, because the `assert!` calls won't provide much information once deployed. See https://soroban.stellar.org/docs/examples/errors.

One larger question I have:
 - It seems like the shares could potentially be a token, which would make the shares transferrable, allowing the owner to migrate the shares from one account to another, or potentially sell them. I imagine this could also be a disadvantage for that latter reason if they aren't intended to be transferred to another person.

This is a really nice example.

**Reply by rahimklaber** [reply]
Thanks! 

I will optimize the contract based or your feedback. For the most part I didn't think about performance. 

Regarding the shares: 
I made the shares non-transfer since I wanted them to work as reputation and I like the idea of a dao where you have to work for your power.

**Reply by leighmcculloch** [reply]
Ah that makes sense. Non transferability is a good feature in this case. It still might be useful developing the shares as a separate token (https://soroban.stellar.org/docs/common-interfaces/token) contract, that just doesn't implement the transfer functions. The advantage would be that wallets would show the token.

**Reply by rahimklaber** [reply]
Hey Leigh, I'm finally looking at this. Is there a reference implementation of the token contract?  Otherwise, A simple solution I thought of is to make the DAO contract the admin of the token and then freeze everyone's account.

**Comment by achallagundla** [comment]
Heya @rahimklaber ,
Congratulations! Your contribution to Sorobanathon: First Light has been accepted! 🎉
Your submission may qualify for a reward of 2500 XLM! One last step: to receive that reward, please fill out this [form](https://stellarform.typeform.com/Sorobanathon) completely and accurately.

Please let us know if you have any questions. (You can also message me on discord @ anuxhya)

All the best,
Anuhya
